src/main/java/htsjdk/samtools/BAMShardIndexer.java

```diff
@@ -0,0 +1,190 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.util.Log;
+
+import java.io.File;
+import java.io.OutputStream;
+
+public class BAMShardIndexer {
+  // output written as binary, or (for debugging) as text
+  private final BinaryBAMShardIndexWriter outputWriter;
+
+  // content is built up from the input bam file using this
+  private final BAMIndexBuilder indexBuilder;
+  
+  // Index of the reference for which the index is being written
+  int reference;
+
+  public BAMShardIndexer(OutputStream output, SAMFileHeader header, int reference) {
+    indexBuilder = new BAMIndexBuilder(header.getSequenceDictionary(), reference);
+    final boolean isFirstIndexShard = reference == 0;
+    final int numReferencesToWriteInTheHeader = isFirstIndexShard ? 
+        header.getSequenceDictionary().size() : 0;
+    outputWriter = new BinaryBAMShardIndexWriter(numReferencesToWriteInTheHeader, output);
+    this.reference = reference;
+  }
+  
+  public void processAlignment(final SAMRecord rec) {
+      try {
+          indexBuilder.processAlignment(rec);
+      } catch (final Exception e) {
+          throw new SAMException("Exception creating BAM index for record " + rec, e);
+      }
+  }
+
+  /**
+   * Finalizes writing and closes the file.
+   * @return count of records with no coordinates.
+   */
+  public long finish() {
+    final BAMIndexContent content = indexBuilder.processReference(reference);
+    outputWriter.writeReference(content);
+    outputWriter.close();
+    return indexBuilder.getNoCoordinateRecordCount();
+  }
+
+  /**
+   * Class for constructing BAM index files.
+   * One instance is used to construct an entire index.
+   * processAlignment is called for each alignment until a new reference is encountered, then
+   * processReference is called when all records for the reference have been processed.
+   */
+  private class BAMIndexBuilder {
+
+      private final SAMSequenceDictionary sequenceDictionary;
+
+      private BinningIndexBuilder binningIndexBuilder;
+
+      private int currentReference = -1;
+
+      // information in meta data
+      private final BAMIndexMetaData indexStats = new BAMIndexMetaData();
+
+      BAMIndexBuilder(final SAMSequenceDictionary sequenceDictionary, int reference) {
+          this.sequenceDictionary = sequenceDictionary;
+          if (!sequenceDictionary.isEmpty()) startNewReference(reference);
+      }
+
+      /**
+       * Record any index information for a given BAM record
+       *
+       * @param rec The BAM record. Requires rec.getFileSource() is non-null.
+       */
+      public void processAlignment(final SAMRecord rec) {
+
+          // metadata
+          indexStats.recordMetaData(rec);
+
+          if (rec.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+              return; // do nothing for records without coordinates, but count them
```

Comment:
Do you mean you count them via the `createIndex` function?  If so, then I recommend adding a bit more to this so folks make the connection in the code.

```Java
for (SAMRecord rec : reader) {
          if (++totalRecords % 1000000 == 0) {
              if (null != log) log.info(totalRecords + " reads processed ...");
          }
          indexer.processAlignment(rec);
      }
```